File: src/main/java/xyz/phanta/cbtweaker/structure/VisualizationGui.java
package xyz.phanta.cbtweaker.structure;

import io.github.phantamanta44.libnine.client.gui.L9Gui;
import io.github.phantamanta44.libnine.util.helper.ItemUtils;
import io.github.phantamanta44.libnine.util.render.GuiUtils;
import io.github.phantamanta44.libnine.util.render.RenderUtils;
import net.minecraft.client.gui.Gui;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.resources.I18n;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import org.apache.commons.lang3.tuple.Triple;
import xyz.phanta.cbtweaker.CbtLang;
import xyz.phanta.cbtweaker.CbtMod;
import xyz.phanta.cbtweaker.gui.CbtTextureResources;
import xyz.phanta.cbtweaker.multiblock.MultiBlockType;
import xyz.phanta.cbtweaker.network.CPacketVisualizationLevel;

import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class VisualizationGui extends L9Gui {

    private static final int VIS_OFF_X = 5, VIS_OFF_Y = 5;

    @Nullable
    private Triple<VisualizationToolItem.VisualizationInfo, ItemStack, VisualizationRenderer> visualization;

    public VisualizationGui() {
        super(CbtTextureResources.GUI_MB_VIS_BG.getTexture().getTexture(),
                CbtTextureResources.GUI_MB_VIS_BG.getWidth(), CbtTextureResources.GUI_MB_VIS_BG.getHeight());
    }

    @Override
    public void initGui() {
        super.initGui();
        updateScreen();
    }

    @Override
    public void drawBackground(float partialTicks, int mX, int mY) {
        super.drawBackground(partialTicks, mX, mY);
        CbtTextureResources.GUI_MB_VIS.draw(getOffsetX() + VIS_OFF_X, getOffsetY() + VIS_OFF_Y);
    }

    @Override
    public void drawForeground(float partialTicks, int mX, int mY) {
        if (visualization == null) {
            GlStateManager.pushMatrix();
            GlStateManager.scale(0.75F, 0.75F, 0.75F);
            drawCenteredString(fontRenderer, I18n.format(CbtLang.GUI_VIS_TOOL_NOT_BOUND),
                    sizeX * 2 / 3, (VIS_OFF_Y + 44) * 4 / 3, 0xFF0000);
            GlStateManager.popMatrix();
            return;
        }

        int visMouseX = mX - VIS_OFF_X, visMouseY = mY - VIS_OFF_Y;
        VisualizationRenderer visRenderer = visualization.getRight();
        visRenderer.handleMouseMovement(visMouseX, visMouseY);
        GlStateManager.enableDepth();
        visRenderer.render(VIS_OFF_X, VIS_OFF_Y, visMouseX, visMouseY);
        RenderUtils.renderItemIntoGui(6, 98, visualization.getMiddle());
        GlStateManager.disableDepth();
        if (isOverControllerSlot(mX, mY)) {
            Gui.drawRect(6, 98, 22, 114, 0x80FFFFFF);
            GlStateManager.color(1F, 1F, 1F, 1F);
        }

        GlStateManager.pushMatrix();
        GlStateManager.scale(0.75F, 0.75F, 0.75F);
        drawString(I18n.format(visualization.getLeft().getMultiBlockType().getTranslationKey()), 35, 138, 0x404040);
        GlStateManager.popMatrix();
    }

    @Override
    public void drawOverlay(float partialTicks, int mX, int mY) {
        if (visualization == null) {
            return;
        }
        List<String> tooltip = new ArrayList<>();
        if (isOverControllerSlot(mX, mY)) {
            ItemUtils.getStackTooltip(visualization.getMiddle(), tooltip, RenderUtils.getTooltipFlags());
        } else {
            visualization.getRight().getTooltip(
                    tooltip, mX - VIS_OFF_X, mY - VIS_OFF_Y, RenderUtils.getTooltipFlags());
        }
        if (!tooltip.isEmpty()) {
            drawTooltip(tooltip, mX, mY);
        }
    }

    @Override
    public boolean doesGuiPauseGame() {
        return false;
    }

    @Override
    public void updateScreen() {
        EntityPlayer player = mc.player;
        VisualizationToolItem.VisualizationInfo newVisInfo = VisualizationToolItem.getVisualizationInfo(player);
        if (newVisInfo == null) {
            visualization = null;
            return;
        }
        MultiBlockType<?, ?, ?> mbType = newVisInfo.getMultiBlockType();
        if (visualization == null || mbType != visualization.getLeft().getMultiBlockType()) {
            VisualizationRenderer visRenderer = new VisualizationRenderer(mbType.getStructureMatcher());
            visRenderer.setLevel(VisualizationToolItem.getLevel(newVisInfo.getStack()));
            visualization = Triple.of(newVisInfo, new ItemStack(mbType.getControllerBlock()), visRenderer);
        } else {
            visualization = Triple.of(newVisInfo, visualization.getMiddle(), visualization.getRight());
        }
    }

    @Override
    protected void mouseClicked(int mX, int mY, int button) {
        if (visualization == null) {
            return;
        }
        VisualizationRenderer visRenderer = visualization.getRight();
        if (visRenderer.handleClick(mX - getOffsetX() - VIS_OFF_X, mY - getOffsetY() - VIS_OFF_Y, button)) {
            VisualizationToolItem.VisualizationInfo visInfo = visualization.getLeft();
            Integer newLevel = visRenderer.getLevel();
            if (!Objects.equals(VisualizationToolItem.getLevel(visInfo.getStack()), newLevel)) {
                CbtMod.INSTANCE.getNetworkHandler()
                        .sendToServer(new CPacketVisualizationLevel(visInfo.getHand(), newLevel));
            }
        }
    }

    @Override
    protected void keyTyped(char typed, int keyCode) {
        if (keyCode == 1 || mc.gameSettings.keyBindInventory.isActiveAndMatches(keyCode)) {
            mc.displayGuiScreen(null);
        }
    }

    private static boolean isOverControllerSlot(int mX, int mY) {
        return GuiUtils.isMouseOver(6, 98, 16, 16, mX, mY);
    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/structure/VisualizationGui.kt
package st.evening.mc.cbtweaker.structure

import net.minecraft.client.gui.GuiScreen
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.client.resources.I18n
import st.evening.mc.cbtweaker.CbTweaker
import st.evening.mc.cbtweaker.CbtLang
import st.evening.mc.cbtweaker.gui.CbtGuiResources
import st.evening.mc.cbtweaker.network.C2SVisualizationLevel
import st.evening.mc.prelude.api.gui.drawable.drawFullSize
import st.evening.mc.prelude.api.util.game.ClientHelper
import st.evening.mc.prelude.api.util.game.ClientSide
import st.evening.mc.prelude.api.util.game.getColouredTooltip
import st.evening.mc.prelude.api.util.render.RenderingHelper
import st.evening.mc.prelude.api.util.render.gui.GuiRenderHelper

@ClientSide.Strong
class VisualizationGui : GuiScreen() {
    companion object {
        private const val VIS_OFF_X: Int = 5
        private const val VIS_OFF_Y: Int = 5

        private fun isOverControllerSlot(mouseX: Int, mouseY: Int): Boolean =
            mouseX >= 1 && mouseY >= 93 && mouseX < 17 && mouseY < 109
    }

    private var windowX: Int = 0
    private var windowY: Int = 0

    override fun initGui() {
        windowX = (width - CbtGuiResources.GUI_MB_VIS_BG.width) / 2
        windowY = (height - CbtGuiResources.GUI_MB_VIS_BG.height) / 2
    }

    override fun drawScreen(mouseX: Int, mouseY: Int, partialTicks: Float) {
        val visX = windowX + VIS_OFF_X
        val visY = windowY + VIS_OFF_Y
        val visMouseX = mouseX - visX
        val visMouseY = mouseY - visY
        val slotX = visX + 1
        val slotY = visY + 93

        drawDefaultBackground()
        CbtGuiResources.GUI_MB_VIS_BG.drawFullSize(partialTicks, windowX, windowY)
        CbtGuiResources.GUI_MB_VIS.drawFullSize(partialTicks, visX, visY)

        val state = ClientVisualizationState.getState()
        if (state == null) {
            RenderingHelper.pushMatrix {
                GlStateManager.scale(0.75F, 0.75F, 0.75F)
                drawCenteredString(
                    fontRenderer,
                    I18n.format(CbtLang.GUI_VIS_TOOL_NOT_BOUND),
                    width * 2 / 3,
                    (windowY + VIS_OFF_Y + 44) * 4 / 3,
                    0xFF0000
                )
            }
        } else {
            state.renderer.handleMouseMovement(visMouseX, visMouseY)

            GlStateManager.enableDepth()
            state.renderer.render(visX, visY, visMouseX, visMouseY)

            RenderingHelper.pushMatrix {
                GlStateManager.scale(0.75F, 0.75F, 0.75F)
                drawString(fontRenderer, I18n.format(state.mbType.translationKey), visX + 35, visY + 138, 0x404040)
            }
            GuiRenderHelper.drawItemAndOverlay(slotX, slotY, state.mbStack)
            GlStateManager.disableDepth()
        }

        val mouseOverSlot = isOverControllerSlot(visMouseX, visMouseY)
        if (mouseOverSlot) {
            GlStateManager.enableBlend()
            GlStateManager.disableTexture2D()
            GlStateManager.color(1F, 1F, 1F, 0.5F)
            GuiRenderHelper.drawUntexturedQuad(slotX, slotY, slotX + 16, slotY + 16)
            RenderingHelper.resetColour()
            GlStateManager.enableTexture2D()
        }

        val tooltip = mutableListOf<String>()
        if (mouseOverSlot) {
            state?.mbStack?.getColouredTooltip(tooltip, ClientHelper.tooltipFlags)
        } else if (state != null) {
            state.renderer.getTooltip(tooltip, visMouseX, visMouseY, ClientHelper.tooltipFlags)
        } else {
            VisualizationRenderer.getUiTooltip(tooltip, visMouseX, visMouseY)
        }
        if (tooltip.isNotEmpty()) {
            drawHoveringText(tooltip, mouseX, mouseY)
        }
    }

    override fun doesGuiPauseGame(): Boolean = false

    protected override fun mouseClicked(mX: Int, mY: Int, button: Int) {
        val state = ClientVisualizationState.getState() ?: return
        if (state.renderer.handleClick(mX - windowX - VIS_OFF_X, mY - windowY - VIS_OFF_Y, button)) {
            val newLevel = state.renderer.level
            if (VisualizationToolItem.getLevel(state.visToolStack) != newLevel) {
                CbTweaker.defns.c2sVisualizationLevel.sendToServer(C2SVisualizationLevel(state.hand, newLevel))
            }
        }
    }

    protected override fun keyTyped(typed: Char, keyCode: Int) {
        if (keyCode == 1 || mc.gameSettings.keyBindInventory.isActiveAndMatches(keyCode)) {
            mc.displayGuiScreen(null)
        }
    }
}
